src/test-table.md

```diff
@@ -100,6 +104,89 @@ for _, tt := range tests {
 }
 ```
 
+### Avoid Unnecessary Complexity in Table Tests
+
+Table tests can be difficult to read and maintain if the subtests contain conditional 
+assertions or other branching logic. Table tests should **NOT** be used whenever 
+there needs to be complex or conditional logic inside subtests (i.e. logic inside the `for` loop).
+
+Large, complex table tests harm readability and maintainability because test readers may
+have difficulty debugging test failures that occur.
+
+Table tests like this should be split into either multiple test tables or multiple 
+individual `Test...` functions.
+
+Some ideals to aim for are:
+
+* Focus on the narrowest unit of behavior
+* Minimize `if` statements and other control flow (if possible)
+* Ensure that all table fields are used in all tests
+* Ensure that all test logic runs for all table cases
+
+Concretely, that means that you should avoid
+having multiple branching pathways (e.g. `shouldError`, `expectCall`, etc.),
+using `if` statements for specific mock expectations (e.g. `shouldCallFoo`),
+or placing functions inside the table (e.g. `setupMocks func(*FooMock)`).
+
+As an exception, if the test body is short and straightforward,
+it's acceptable to have a single branching pathway for success versus failure cases
+with a table field like `wantErr` to specify error expectations.
+
+For example, the following table test is confusing to read:
+
+_Bad:_
+```go
+func TestComplicatedTable(t *testing.T) {
+	tests := []struct {
+		give          string
+		want          string
+		wantErr       error
+		shouldCallX   bool
+		shouldCallY   bool
+		giveXResponse string
+		giveXErr      error
+		giveYResponse string
+		giveYErr      error
+	}{
+		// ...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.give, func(t *testing.T) {
+			// setup mocks
+			ctrl := gomock.NewController(t)
+			xMock := xmock.NewMockX(ctrl)
+			yMock := xmock.NewMockY(ctrl)
+
+			// run the test
+			got, err := DoComplexThing(tt.give, xMock, yMock)
+
+			// verify results
+			if tt.wantErr != nil {
+				require.EqualError(t, err, tt.wantErr)
+				return
+			}
+			require.NoError(t, err)
+			if tt.shouldCallX {
+				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
+			}
+			if tt.shouldCallY {
+				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
+			}
```

Comment:
These EXPECT calls should also be in the "setup mocks" section.
gomock (which is what this looks like) wants expectations set in advance.
```suggestion
			yMock := xmock.NewMockY(ctrl)
			if tt.shouldCallX {
				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
			}
			if tt.shouldCallY {
				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
			}

			got, err := DoComplexThing(tt.give, xMock, yMock)

			// verify results
			if tt.wantErr != nil {
				require.ErrorIs(t, err, tt.wantErr)
				return
			}
			require.NoError(t, err)
```